### DISCUSSION TITLE
question on storage tiering on speed with classes

### DISCUSSION THREAD

**Discussion by eleaner** [moosefs/moosefs#598] [General]
My main usage of moosefs is storage of photos.
I work on them and then move to archive
my WIP folder class is FAST(3SSD) while ARCH is SLOW(3HDD)
I snapshot both folders regularly

so imagine my workflow
- it took some photos -> store them in WIP (FAST)
- (snapshots)
- i work on them in WIP (FAST)
- (snapshots)
- if finished -> move to ARCH and change the class to SLOW 

in this moment the same chunks that previously were snapshotted with class FAST receive also class SLOW and moosefs falls to class 3* - this way my ARCHIVE uses my precious SSD space.

how can I manage it?
I tried changing my snapshots to SLOW but that's even worse
with first snapshot my WIP photos are moved to 3* because WIP is FAST and snapshot of WIP is SLOW.

What are my options?
I guess in theory I could write a script that:
first ensures that WIP is FAST and ARCH is SLOW
and then traverses all snapshots searching for the same files in WIP and ARCH (I guess by name) and sets them to the same class with default SLOW for the files that are deleted

That's possible but has two problems
1. it blows when the file name has been changed
2. with my amount of photos the process will take hours 

another option would be to change definition of class 3* (or where moosefs falls to) to the same as FAST
then assuming WIP = FAST, ARCH = SLOW and snapshots = SLOW
the only photos in fallback will be the ones in WIP

but can I redefine 3*?

**Comment by eleaner** [comment]
@chogata @xandrus 
I brought my unresolved support query here because I guess it is generic enough

**Comment by eleaner** [comment]
I wrote a little script,
runs over every photos, checks its class and stores in an array (that's in attempt to to improve performance)
~700k files takes 3.5h
I have 35 snapshots at any given time
I am going to compare class of each photo there with the array
That means running mfsgetsclass for each and aligning them is going to take roughly 35*3.5h (based on the above)

It would mean that to align the classes with this method (assuming it works) it will take about a week.

That's a bit excessive and borderline insane.
This would mean that I am likely not seeing something obvious.
I cannot be the only person in the world with such requirement.

**Comment by jkiebzak** [comment]
When you move a file from wip to archive, could you change the sclass at the same time for that file? That way you don’t need to walk the folder structure

**Comment by eleaner** [comment]
@jkiebzak 
that's precisely the problem

i do change it 
but then it is SLOW in archive and FAST in all previous snapshots 
when classes on one chunk are in conflict moosefs falls back to standard classes = will actually use 3*

**Comment by eleaner** [comment]
I think my script could be multiple times faster if mfsgetsclass could return a list of files (with complete path) and their class, recursively
kind of like option -r but not summary

**Comment by eleaner** [comment]
well, or if mfssetsclass had an option to synchronise the class across all paths

**Comment by eleaner** [comment]
@onlyjob 
sorry for pulling you here but you always have good ideas

**Comment by eleaner** [comment]
ok, so one performance kick is `find` running on millions of files
I was trying to use `mfsfilepaths` instead but it never shows me snapshots, how should I use it?

**Reply by chogata** [reply]
[`mfsfilepaths`](https://moosefs.com/manpages/mfsfilepaths) is a tool to show all paths of the SAME file (same inode – basically hard links). Snapshots are different files (different inodes), that just share chunks. A file, that is a snapshot, doesn't "know" from which file it was snapshotted. If you want to connect snapshots to original files, it is only possible if they share at least one chunk. How to do it:
- list all chunks of a snapshotted file ([`mfsfileinfo`](https://moosefs.com/manpages/mfsfileinfo))
- use [`mfsmetasearch`](https://moosefs.com/manpages/mfsmetasearch) to find all files that have these chunks – from these, the one that doesn't have the snapshot flag is your "original" (unless you deleted it, for example)
If a snapshotted file and an original file no longer share even one chunk (that means, all the chunks were modified at some point after the snapshot happened), there is no way to tell there is any connection between them.
